### PR TITLE
[gpt_oss_d_p] chunked prefill: SP ring cache-read (sinks on all layers)

### PR DESCRIPTION
**Pull request by mdragulaTT**
> **Chunked prefill — the last piece on top of the merged stack.** #50223 (attention) + #50228 (KV cache) + #50254 (MoE) + #50265 (runtime) are merged, and #51438 (Pavle's `RingJointSDPA` sliding+sink op) is merged. This wires the SP ring cache-read on top, enabling multi-chunk prefill.

## What
Enables **multi-chunk prefill** for GPT-OSS on the galaxy (SP=4 × TP=8). Each chunk after the first attends the accumulated KV prefix (cache-read) instead of re-processing the whole prompt one-shot. This is the **disaggregation gate** — KV migration can't be tested until chunked works. Chunk 0 keeps the gather-Q stand-in; chunks 1+ use the native ring-joint dense SDPA (#51438) over the block-cyclic SP-sharded KV cache.

## This PR
- `tt/attention/dense_sp.py` (**new**): ring cache-read helper over the block-cyclic packed KV cache via `ttnn.transformer.ring_joint_scaled_dot_product_attention`. Compact halo gather for sliding layers (`max(ceil((window-1)/k_chunk)*k_chunk, TILE)`), full-sequence gather for full-attention layers; bf8 K/V; `is_balanced=False` (chunked); `kv_cache_batch_idx = slot*num_layers + layer`.
- `tt/attention/prefill.py`: the SP seam now wires **chunks 1+** (`cached_len>0`) to `dense_sp_attention` (was `NotImplementedError`). Chunk 0 (`cached_len==0`) keeps the gather-Q stand-in — the sliding op requires `is_chunked` (Q shorter than the cache), which chunk 0 is not. **Sinks on ALL layers:** the ring op folds the sink once across ring iterations, so both sliding-window and full-attention layers are sink-correct (`fp32_dest_acc_en=False` for the streaming sink compute).
- `tt/tt_prefill_runtime.py` + `tests/galaxy_prefill_kv_pcc.py`: drop the stale `chunked = NotImplementedError` notes/warnings; the harness chunked driver (already present from #50265) now runs.

## Validation status (read before reviewing)
Galaxy-validated on **this exact branch built from main** (single 4×8 Blackhole galaxy, real bf16 weights, 36L @ 5k):
- ✅ **Chunked (2×2560) runs end-to-end → min KV PCC 0.823** vs **one-shot 0.859** on the same branch. Same smooth per-layer decline, no discontinuity → the ring cache-read with sinks is numerically correct; the ~0.036 chunked-vs-one-shot gap is the ring cache-read numerical drift (bf8 gather + online softmax). Both sit at the bf8-compute floor.
- ✅ Full `build_metal.sh` from main + `ttnn` import from the branch; the chunked gate exercises chunk 0 (gather-Q) + chunks 1+ (ring cache-read) across all 36 layers.
- ✅ **55k chunked (27 chunks)** ran end-to-end on an earlier base (ring op scales to ~11× the 5k length); not re-run on this branch.
- ℹ️ Peak PCC is not the accuracy bar here — it's the bf8-compute floor (see the accuracy investigation). The functional sign-off is **top-1 / logits agreement**, which lands as a separate follow-up PR.

## Known limits carried forward (tracked)
- **Single-chip (sp==1) cache-read** is still `NotImplementedError` (not used on the galaxy; SP>1 is the path).
- **top-1 functional sign-off** + the **`ATTN_WEIGHT_DTYPE`** diagnostic knob are intentionally **not** in this PR (separate follow-up).
- **Perf** *(corrected — see #52000)*: chunked **warm steady-state is ~1.28s at 5k (~2x one-shot 0.64s)** — the ring cache-read's real overhead. The earlier "~16x" was a measurement artifact: a one-time empty on-disk kernel-cache compile (~10.6s, never recurs) plus `compile()` not warming the ring path, so its kernels JIT-compiled inside the single timed chunk. The `compile()` warmup fix lands in #52022 (first served chunk 2.16s -> 1.26s); profiling/tuning the residual ~2x is tracked in #52000. Long-context still requires chunked (one-shot OOMs at 55k).

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:mdragula/gpt-oss-chunked-prefill)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:mdragula/gpt-oss-chunked-prefill)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:mdragula/gpt-oss-chunked-prefill)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:mdragula/gpt-oss-chunked-prefill)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:mdragula/gpt-oss-chunked-prefill)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:mdragula/gpt-oss-chunked-prefill)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:mdragula/gpt-oss-chunked-prefill)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:mdragula/gpt-oss-chunked-prefill)

### CODE REVIEW

File: models/demos/gpt_oss_d_p/tt/attention/prefill.py [hunk 1]
```diff
@@ -177,64 +178,83 @@ def attention_forward(
...
-            memory_config=ttnn.DRAM_MEMORY_CONFIG,
-            topology=ccl_manager.topology,
-            cluster_axis=mesh_config.sp_axis,
-            barrier_semaphore=ccl_manager.get_barrier_semaphore(),
-        )
-        tt_sdpa_out_full.deallocate(True)
-        tt_sdpa_out_scaled = ttnn.multiply(tt_sdpa_out, 1.0 / sp)
-        ttnn.deallocate(tt_sdpa_out)  # free the reduce-scatter output (leaked otherwise across 36 layers)
-        tt_sdpa_out = tt_sdpa_out_scaled
+            sp_kcfg = ttnn.WormholeComputeKernelConfig(
```
Comment (jmalone-tt) [inline models/demos/gpt_oss_d_p/tt/attention/prefill.py]: does this work for BH too?

**Review by jmalone-tt — APPROVED** [review]
looks good other than one comment (please also address copilot comments)

File: models/demos/gpt_oss_d_p/tt/attention/prefill.py (hunk 1 above)
Comment (mdragulaTT) [reply models/demos/gpt_oss_d_p/tt/attention/prefill.py]: Good catch — there's currently no Blackhole `ComputeKernelConfig`, so I used the Wormhole one, but I agree it's misleading. The chunked prefill *was* tested on BH. I switched to the device-agnostic `init_device_compute_kernel_config(mesh_device.arch(), ...)` to be less misleading, and opened #51998 to add a Blackhole config someday.

File: models/demos/gpt_oss_d_p/tt/attention/prefill.py [hunk 2]
```diff
@@ -177,64 +178,83 @@ def attention_forward(
...
+                kv_actual=cached_len,
+                logical_n=cached_len + seq_len * sp,
+                n_kv=config.num_kv_heads,
+                cache_global=kv_cache.max_seq_len,
+                head_dim=config.head_dim,
+                mesh_device=mesh_device,
+                ccl_manager=ccl_manager,
+                program_config=sp_prog,
+                compute_kernel_config=sp_kcfg,
+                scale=config.head_dim**-0.5,
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply models/demos/gpt_oss_d_p/tt/attention/prefill.py]: Done — now `scale=config.scaling`, consistent with the sink pre-division (equals `head_dim**-0.5` for the default config; correct if `scaling` is ever set).

File: models/demos/gpt_oss_d_p/tt/attention/dense_sp.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        cache_k,
+        cache_v,
+        None,
+        None,
+        None,
+        # Persistent ring-gather scratch (once per key/size, reused across layers/chunks). Full-attn
+        # layers gather the entire per-device shard (cache_global); sliding layers use a compact halo
+        # (_bufseq). logical_n / kv_actual_isl drive causal masking. dtype MUST match the bf8 KV cache.
+        persistent_output_buffer_k=ccl_manager.get_ring_gather_buffer(
+            f"dense_k_{_bufseq}", n_kv, _bufseq, head_dim, ttnn.bfloat8_b
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply models/demos/gpt_oss_d_p/tt/attention/dense_sp.py:119]: Done — asserts bf8 K/V up front with a clear message; `KV_CACHE_DTYPE=bf16` is unsupported for chunked, so it fails early instead of on chunk 2.

File: models/demos/gpt_oss_d_p/tt/attention/prefill.py [hunk 4]
```diff
@@ -177,64 +178,83 @@ def attention_forward(
...
-        ttnn.deallocate(tt_sdpa_out)  # free the reduce-scatter output (leaked otherwise across 36 layers)
-        tt_sdpa_out = tt_sdpa_out_scaled
+            sp_kcfg = ttnn.WormholeComputeKernelConfig(
+                math_fidelity=ttnn.MathFidelity.HiFi4,
+                math_approx_mode=False,
+                fp32_dest_acc_en=False,  # ring op requires streaming compute for attention_sink
+                packer_l1_acc=False,
+            )
+            assert kv_cache is not None, "SP chunked cache-read needs a KV cache"
+            tt_sdpa_out = dense_sp_attention(
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply models/demos/gpt_oss_d_p/tt/attention/prefill.py:205]: Agreed — the chunked mode of `tests/galaxy_prefill_kv_pcc.py` is the integration test but isn't wired into the tiered pipeline yet (galaxy-gated). Tracking it as a follow-up; not blocking this bring-up PR.

File: models/demos/gpt_oss_d_p/tt/attention/dense_sp.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+  * :func:`dense_sp_attention` — cache-read (``cached_len > 0``, multi-chunk): ring_joint over the
+    accumulated prefix in the block-cyclic SP KV cache. This is the path that unlocks CHUNKED prefill.
+    Chunk 0 (``cached_len == 0``) keeps the AllGather-Q + single-chip SDPA stand-in (equal Q/cache
+    shapes, so the ring op's is_chunked requirement does not apply).
+
+GPT-OSS vs M3: gpt-oss attention has **attention sinks** and **per-layer sliding-window** masking, so
+both paths thread ``attention_sink`` + ``sliding_window_size`` into the op (M3 has neither). These are
+Pavle's ``ring-joint-sdpa-attention-sinks`` additions (unmerged — needs his branch cherry-picked +
+rebuilt to run). ``attention_sink`` is our ``weights.sinks`` (stored ``sinks_div_scale`` = sink×√d, the
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply models/demos/gpt_oss_d_p/tt/attention/dense_sp.py]: Done — dropped the stale unmerged/cherry-pick note; the ring op is merged (#51438) and the code calls the merged API.
